File: backend/app/repositories/topic_repo.py
from sqlalchemy.orm import Session
from typing import List, Optional
from app.models.topic import Topic

class TopicRepository:
    def __init__(self, db: Session):
        self.db = db

    def get_by_id(self, topic_id: int) -> Optional[Topic]:
        return self.db.query(Topic).filter(Topic.id == topic_id).first()

    def get_all(self) -> List[Topic]:
        return self.db.query(Topic).order_by(Topic.name).all()

    def create(self, topic: Topic) -> Topic:
        self.db.add(topic)
        self.db.commit()
        self.db.refresh(topic)
        return topic

    def delete(self, topic: Topic) -> None:
        self.db.delete(topic)
        self.db.commit()
        
    def save(self) -> None:
        self.db.commit()

File: backend/app/repositories/topic_repo.js
const db = require('../database/session');

class TopicRepository {
  static async getById(id) {
    const col = await db.collection('topics');
    return await col.findOne({ _id: id }) || null;
  }

  static async getAll() {
    const col = await db.collection('topics');
    return await col.find({}).sort({ name: 1 }).toArray();
  }

  static async create(topic) {
    const col = await db.collection('topics');
    const id = await db.getNextSequenceValue('topic_id');
    const doc = {
      _id: id,
      id: id,
      name: topic.name,
      description: topic.description,
      created_at: new Date()
    };
    await col.insertOne(doc);
    return doc;
  }

  static async delete(id) {
    const col = await db.collection('topics');
    const res = await col.deleteOne({ _id: id });
    return res.deletedCount > 0;
  }
}

module.exports = TopicRepository;
